rts/System/SpringMem.h

```diff
@@ -0,0 +1,40 @@
+/* This file is part of the Spring engine (GPL v2 or later), see LICENSE.html */
+
+#ifndef SPRING_MEM_H
+#define SPRING_MEM_H
+
+#include <vector>
+#include <map>
+
+namespace spring {
+	void* AllocateAlignedMemory(size_t size, size_t alignment);
+	void FreeAlignedMemory(void* ptr);
+}
+
+namespace spring {
+    template <typename T>
+    class StablePosAllocator {
```

Comment:
Oh yes, it creeped in from one feature branch.

In essence the current implementation of MatrixUploader is less than ideal because the offset into SSBO array is not guaranteed to persist across draw frames, which makes it less usable than should from both engine (I have another half-done implementation that sources matrices for engine drawing completely from SSBO) and from Lua.

The future implementation (half done) will use this allocator to request a slot for as many unsynced matrices as needed.
For example for unit it will be world matrix + all piece matrices and will keep this slot until unit is dead. Upon death the slot will become free and either several smaller slots can be allocated in the free place or similar new unit can occupy it. The array from allocator will be uploaded to GPU's SSBO without any extra mess. Some stuff is done here: https://github.com/beyond-all-reason/spring/tree/BAR-matricesAllocator but I'm less than happy with overall implementation so it might get scrapped. The allocator itself proved to be robust in my "give all"/kill all and AI tests.